### DISCUSSION TITLE
How can it be updated to Flipper API v0.71?

### DISCUSSION THREAD

**Discussion by unrealsolver** [dcoles/flipperzero-rs#23] [Q&A] [answered]

**Comment by dcoles** [comment, marked as answer]
Hi @unrealsolver,

The current [0.4.1 release](https://crates.io/crates/flipperzero-sys/0.4.1) should support SDK version 7.5 ([flipperzero-firmware@0.71.1](https://github.com/flipperdevices/flipperzero-firmware/tree/0.71.1)).

Let me know if you're running into an issue on that version of the firmware.

**Comment by unrealsolver** [comment]
I mean, how can I specify which firmare to use? I grep'd through the code base, but there are no explicit versions mentioned (like "0.68") or any scripts/external dependecies, which download flipperzero-firmware from github. So I don't really understand how does flipperzero-rs is connected to appropriate version of flipperzero-firmware.

**Reply by dcoles** [reply]
The way I currently handle this is to use [`generate-bindings`](https://github.com/dcoles/flipperzero-rs/blob/main/tools/src/bin/generate-bindings.rs) (e.g. 9daed098d1a09e3a0885e922a28d1001094f20b9) to regenerate [`bindings.rs`](https://github.com/dcoles/flipperzero-rs/blob/main/crates/sys/src/bindings.rs) when the SDK becomes incompatible. It uses [bindgen](https://crates.io/crates/bindgen) internally.

You can see the latest supported SDK version on [`flipperzero-sys`](https://crates.io/crates/flipperzero-sys), but I'll see if I can make it more visible in the README/documentation.

The reason I take this approach, rather than using a `build.rs` script is that currently the Flipper Zero SDK changes almost every release. You wouldn't be able to reliably build bindings without pinning it to a specific `flipperzero-firmware` release and it would add a build-time dependency on `flipperzero-firmware` and its toolchain.

If you ever notice that it's out of date, feel free to open an issue and I'll happily tag a new release.

**Reply by unrealsolver** [reply]
Does this process require manual labour? Or it could be just run script and build process? Because it might be simpler to set some env var (like in the example) and regenerate bindings automatically.

**Reply by dcoles** [reply]
Unfortunately, yes. I have typically needed to make non-trivial changes to `generate-bindings` to support each release.

Additionally, the external application SDK has broken backwards compatibility several times.
As an example, the 0.71.1 release has [SDK version 7.5](https://github.com/flipperdevices/flipperzero-firmware/blob/0.71.1/firmware/targets/f7/api_symbols.csv), while the 0.73.1 release has [SDK version 10.1](https://github.com/flipperdevices/flipperzero-firmware/blob/0.73.1/firmware/targets/f7/api_symbols.csv).

This means that any crate directly depending on [`flipperzero-sys`](https://crates.io/crates/flipperzero) (including [`flipperzero`](https://crates.io/crates/flipperzero) and [`flipperzero-rt`](https://crates.io/crates/flipperzero-rt)) will likely need code changes to handle the API breakage.

**Comment by GitCat3** [comment]
yeah I just got the api mismatch error on my dev firmware, is it only updated to release?

**Reply by dcoles** [reply]
Yes. This would become `0.5.x`.

I'll take a look into updating it today.

**Reply by GitCat3** [reply]
I am trying to look through the generate bindings thing, and at the top where you are defining `TOOLCHAIN`, where are you getting that toolchain from?

**Reply by dcoles** [reply]
Once you have a checkout of [flipperzero-firmware](https://github.com/flipperdevices/flipperzero-firmware/), run `./fbt` at least once. That will ensure that the toolchain is downloaded (into `toolchain` folder).

**Reply by dcoles** [reply]
The [0.5.0 release](https://crates.io/crates/flipperzero/0.5.0) now supports Flipper Zero API version 10.1.

**Reply by GitCat3** [reply]
one last thing, what is the `sdk.opts` file?

**Reply by GitCat3** [reply]
I think I am confused on where to put the files. I have a flipperzero firmware that I have ran `./fbt` on already, which folder do I need to clone this project to in order to generate bindings?

**Reply by dcoles** [reply]
After running `./fbt` in the flipper-firmware directory, there should now be be a `build/f7-firmware-D/sdk/` directory.

The `sdk.opts` file in this folder give the path of important directories and the compiler flags that bindgen needs to use.

After running `generate-bindings ${FLIPPERZERO_FIRMWARE}/build/f7-firmware-D/sdk/`, there should be a new `bindings.rs` in the current directory. You can then use this to replace the one in `crates/sys/src/`.

**Reply by GitCat3** [reply]
So in order to do that would I run the `cargo run` with the arguments that you showed?

**Reply by dcoles** [reply]
Yes. This was the command I used:
```
cd flipperzero-rs/tools
cargo run --bin generate-bindings ../../flipperzero-firmware/build/f7-firmware-D/sdk
```
